// @applitools/tunnel-client/dist/cli/cli.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const worker_1 = require("../worker");
const yargs_1 = __importDefault(require("yargs"));
const version_1 = require("../version");
yargs_1.default
    .command({
    command: '*',
    builder: {
        pollingServerUrl: {
            alias: ['baseUrl'],
            description: 'Polling server URL.',
            type: 'string',
            require: true,
        },
        secret: {
            description: 'Auth secret for polling server.',
            type: 'string',
            require: true,
        },
        agentId: {
            description: 'Agent name.',
            type: 'string',
            require: true,
        },
        pacFile: {
            description: 'Absolute or relative path to a PAC file on disk. AUT traffic is routed via the proxy chain the PAC dictates. Env var: APPLITOOLS_EG_TUNNEL_PAC_FILE.',
            type: 'string',
            default: process.env.APPLITOOLS_EG_TUNNEL_PAC_FILE,
        },
        pacScript: {
            description: 'Inline PAC source as a string. Takes precedence over pacUrl; pacFile takes precedence over pacScript. Env var: APPLITOOLS_EG_TUNNEL_PAC_SCRIPT.',
            type: 'string',
            default: process.env.APPLITOOLS_EG_TUNNEL_PAC_SCRIPT,
        },
        pacUrl: {
            description: 'HTTP(S) URL serving a PAC file; fetched once at startup with a 10 s timeout. Env var: APPLITOOLS_EG_TUNNEL_PAC_URL.',
            type: 'string',
            default: process.env.APPLITOOLS_EG_TUNNEL_PAC_URL,
        },
    },
    handler: settings => {
        var _a;
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = (_a = process.env.NODE_TLS_REJECT_UNAUTHORIZED) !== null && _a !== void 0 ? _a : '0';
        (0, worker_1.makeTunnelClientWorker)({ settings });
    },
})
    .version(version_1.VERSION)
    .wrap(yargs_1.default.terminalWidth()).argv;
