// @applitools/eg-socks5-proxy-server/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVENTS = exports.SocksServer = void 0;
exports.createSocks5ProxyServer = createSocks5ProxyServer;
const constants_1 = require("./constants");
const EVENTS = __importStar(require("./events"));
const end_socket_connection_1 = require("./end-socket-connection");
const handshake_1 = require("./handshake");
const authenticate_1 = require("./authenticate");
const parse_connect_buffer_1 = require("./parse-connect-buffer");
const execute_request_1 = require("./execute-request");
const execute_request_through_remote_proxy_1 = require("./execute-request-through-remote-proxy");
const try_http_proxy_chain_1 = require("./try-http-proxy-chain");
const domain_1 = __importDefault(require("domain"));
const net_1 = __importDefault(require("net"));
const is_private_address_1 = require("./utils/is-private-address");
const _alwaysTrue = (data) => true;
/**
 * The following RFCs may be useful as background:
 *
 * https://www.ietf.org/rfc/rfc1928.txt - NO_AUTH SOCKS5
 * https://www.ietf.org/rfc/rfc1929.txt - USERNAME/PASSWORD SOCKS5
 *
 **/
class SocksServer {
    constructor(options) {
        this.activeSessions = [];
        this.destinationSockets = [];
        this.on = (event, listener) => this.server.on(event, listener);
        this.listen = (...args) => this.server.listen(...args);
        this.listenAsync = (port, host = '127.0.0.1') => {
            const promise = new Promise((resolve, reject) => {
                try {
                    this.server.listen(port, host, resolve);
                }
                catch (e) {
                    reject(e);
                }
            });
            return promise;
        };
        this.close = (...args) => {
            this.syncDestinationSocketInterval && clearInterval(this.syncDestinationSocketInterval);
            return this.server.close(...args);
        };
        this.getConnections = (cb) => this.server.getConnections(cb);
        // SERVER-SIDE upstream-SOCKS5 chain (used by eg-chromedriver-image for
        // tunnel load-balancing). Not for client-side corporate-proxy routing —
        // see `setHttpProxyResolver` for that.
        this.setProxyServer = ({ address, port, shouldUseProxy }) => {
            const getAvailableProxy = () => ({ address, port });
            this.options.proxyServers = { shouldUseProxy, getAvailableProxy };
        };
        this.setProxyServers = (proxyServers) => {
            this.options.proxyServers = proxyServers;
        };
        this.getProxyServers = () => this.options.proxyServers;
        // CLIENT-SIDE per-request HTTP CONNECT proxy resolver (used by
        // execution-grid-tunnel to route AUT traffic through a customer's corporate
        // HTTP proxy, typically driven by a PAC-file evaluator). Returning `null`
        // from the resolver lets the request fall through to the SOCKS5-chain
        // dispatcher branch, preserving compatibility with the server-side use.
        this.setHttpProxyResolver = (resolver) => {
            this.options.httpProxyResolver = resolver;
        };
        const self = this;
        this.options = options || {};
        this.syncDestinationSocketInterval = setInterval(() => {
            this.destinationSockets = this.destinationSockets.filter((dest) => !dest.destroyed);
        }, 5000);
        this._server = net_1.default.createServer((socket) => {
            let clientHandshakeBuffer;
            socket.on('error', (err) => {
                self.server.emit(EVENTS.ORIGIN_SOCKET_ERROR, { err, socket });
            });
            socket.setKeepAlive(true);
            function _authenticate(buffer) {
                (0, authenticate_1.authenticate)({
                    socket,
                    buffer,
                    options: self.options,
                    connect,
                    end,
                    onAuthenticated: (args) => self.server.emit(EVENTS.AUTHENTICATION, args.username.toString()),
                    onFailed: (args, err) => self.server.emit(EVENTS.AUTHENTICATION_ERROR, args.username.toString(), err),
                });
            }
            /**
             * +----+-----+-------+------+----------+----------+
             * |VER | CMD |  RSV  | ATYP | DST.ADDR | DST.PORT |
             * +----+-----+-------+------+----------+----------+
             * | 1  |  1  | X'00' |  1   | Variable |    2     |
             * +----+-----+-------+------+----------+----------+
             *
             * @param {Buffer} buffer - a buffer
             * @returns {undefined}
             **/
            function connect(buffer) {
                const onParseCompleted = (args) => {
                    if (args.cmd !== constants_1.RFC_1928_COMMANDS.CONNECT) {
                        // bind and udp associate commands
                        return end(constants_1.RFC_1928_REPLIES.SUCCEEDED, args);
                    }
                    const connectionFilterDomain = domain_1.default.create();
                    let connectionFilter = self.options.connectionFilter;
                    // if no connection filter is provided, stub one
                    if (!connectionFilter || typeof connectionFilter !== 'function') {
                        connectionFilter = (destination, origin, callback) => setImmediate(callback);
                    }
                    // capture connection filter errors
                    connectionFilterDomain.on('error', (err) => {
                        // emit failed destination connection event
                        self.server.emit(EVENTS.CONNECTION_FILTER, 
                        // destination
                        {
                            address: args.dst.addr,
                            port: args.dst.port,
                        }, 
                        // origin
                        {
                            address: socket.remoteAddress,
                            port: socket.remotePort,
                        }, err);
                        // respond with failure
                        return end(constants_1.RFC_1928_REPLIES.CONNECTION_NOT_ALLOWED, args);
                        //return end(RFC_1929_REPLIES.CONNECTION_NOT_ALLOWED, args)
                    });
                    // perform connection
                    return connectionFilter(
                    // destination
                    {
                        address: args.dst.addr,
                        port: args.dst.port,
                    }, 
                    // origin
                    {
                        address: socket.remoteAddress,
                        port: socket.remotePort,
                    }, connectionFilterDomain.intercept(async () => {
                        var _a;
                        const destinationInfo = {
                            address: args.dst.addr,
                            port: args.dst.port,
                        };
                        const originInfo = {
                            address: socket.remoteAddress,
                            port: socket.remotePort,
                        };
                        // whitelisted addrsses are never routed through a proxy (tunnel) and pass all validations
                        // the main use is browser extension communication with internal EC services (e.g. local test runner)
                        // const isWhitelisted =
                        //   self.options.whitelist?.indexOf(
                        //     `${destinationInfo.address}:${destinationInfo.port}`,
                        //   ) !== -1
                        const isWhitelisted = self.options.shouldWhitelist && self.options.shouldWhitelist(destinationInfo);
                        const shouldBlockRequest = !isWhitelisted &&
                            self.options.shouldBlockRequest &&
                            self.options.shouldBlockRequest(destinationInfo);
                        if (shouldBlockRequest) {
                            self.server.emit(EVENTS.ADDRESS_WAS_BLOCKED, {
                                destinationInfo,
                                originInfo,
                            });
                            return end(constants_1.RFC_1928_REPLIES.CONNECTION_NOT_ALLOWED, args);
                        }
                        if (!isWhitelisted &&
                            shouldBlockRequest !== false &&
                            self.options.shouldBlockPrivateAddresses &&
                            (await (0, is_private_address_1.isPrivateAddress)(destinationInfo.address))) {
                            self.server.emit(EVENTS.PRIVATE_ADDRESS_WAS_BLOCKED, {
                                destinationInfo,
                                originInfo,
                            });
                            return end(constants_1.RFC_1928_REPLIES.CONNECTION_NOT_ALLOWED, args);
                        }
                        // === CLIENT-SIDE HTTP CONNECT branch ===
                        //
                        // Consulted BEFORE the SERVER-SIDE SOCKS5-chain branch below.
                        // The resolver decides per-request whether this destination is
                        // reachable only through a customer corporate HTTP proxy
                        // (typical caller: a PAC-file evaluator wrapped in
                        // `@applitools/execution-grid-tunnel`).
                        //
                        // The resolver may return:
                        //   - a single `HttpProxyDecision` (single proxy)
                        //   - a `ProxyChainEntry[]` (ordered PAC fallback chain;
                        //     entries may be `'DIRECT'` to mean PAC's DIRECT directive)
                        //   - `null`/`undefined`/`[]` — fall through to the existing
                        //     SOCKS5-chain branch, preserving server-side behaviour
                        //     (where this resolver is never set).
                        //
                        // Whitelisted destinations always bypass the resolver — same
                        // policy as the existing proxyServers branch.
                        if (!isWhitelisted && self.options.httpProxyResolver) {
                            let resolverResult;
                            try {
                                resolverResult = await self.options.httpProxyResolver({
                                    socket,
                                    destinationInfo,
                                    originInfo,
                                });
                            }
                            catch (resolverErr) {
                                self.server.emit(EVENTS.HTTP_PROXY_CONNECT_FAILED, {
                                    destinationInfo,
                                    originInfo,
                                    proxyInfo: undefined,
                                    error: resolverErr,
                                });
                                return end(constants_1.RFC_1928_REPLIES.GENERAL_FAILURE, args);
                            }
                            const chain = (0, try_http_proxy_chain_1.normalizeChain)(resolverResult);
                            if (chain) {
                                const firstEntry = chain[0];
                                const firstProxyInfo = firstEntry === 'DIRECT'
                                    ? undefined
                                    : { address: firstEntry.address, port: firstEntry.port };
                                self.server.emit(EVENTS.ACCEPT_NEW_REQUEST, {
                                    destinationInfo,
                                    originInfo,
                                    proxyInfo: firstProxyInfo,
                                });
                                // Whole-chain budget timeout (mirrors the existing
                                // per-request timeout). Fires REMOTE_CONNECTION_TIMEOUT_ERROR
                                // observationally — the inner `executeRequest…` modules
                                // have their own tighter timeouts that actually abort.
                                const connectionTimeout = 120000;
                                let connectionTimeoutId = setTimeout(() => self.server.emit(EVENTS.REMOTE_CONNECTION_TIMEOUT_ERROR, {
                                    destinationInfo,
                                    originInfo,
                                    proxyInfo: firstProxyInfo,
                                    timeout: connectionTimeout,
                                }), connectionTimeout);
                                const clearConnectionTimeout = () => {
                                    if (connectionTimeoutId)
                                        clearTimeout(connectionTimeoutId);
                                    connectionTimeoutId = null;
                                };
                                let domainExited = false;
                                const exitDomainOnce = () => {
                                    if (domainExited)
                                        return;
                                    domainExited = true;
                                    connectionFilterDomain.exit();
                                };
                                const successfulDestination = await (0, try_http_proxy_chain_1.tryHttpProxyChain)({
                                    originalSocket: socket,
                                    chain,
                                    addressType: args.atyp,
                                    address: args.dst.addr,
                                    port: args.dst.port,
                                    requestBuffer: args.requestBuffer,
                                    destinationInfo,
                                    originInfo,
                                    end,
                                    registerDestination: (d) => self.destinationSockets.push(d),
                                    callbacks: {
                                        onConnectedToHttpProxy: (entry, dest) => {
                                            exitDomainOnce();
                                            clearConnectionTimeout();
                                            self.server.emit(EVENTS.CONNECTED_TO_HTTP_PROXY, {
                                                destinationInfo,
                                                originInfo,
                                                proxyInfo: { address: entry.address, port: entry.port },
                                            });
                                        },
                                        onConnectedToRemoteAddress: (dest) => {
                                            exitDomainOnce();
                                            clearConnectionTimeout();
                                            self.server.emit(EVENTS.CONNECTED_TO_REMOTE_ADDRESS, {
                                                destinationInfo,
                                                originInfo,
                                                proxyInfo: undefined,
                                            });
                                        },
                                        onProxyResponded: (statusCode, entry) => {
                                            self.server.emit(EVENTS.HTTP_PROXY_CONNECT_RESPONDED, {
                                                destinationInfo,
                                                originInfo,
                                                proxyInfo: { address: entry.address, port: entry.port },
                                                statusCode,
                                            });
                                        },
                                        onProxyFailed: (rfc1928Reply, entry) => {
                                            self.server.emit(EVENTS.HTTP_PROXY_CONNECT_FAILED, {
                                                destinationInfo,
                                                originInfo,
                                                proxyInfo: { address: entry.address, port: entry.port },
                                                rfc1928Reply,
                                            });
                                        },
                                        onSocketError: (err, entry, dest) => {
                                            exitDomainOnce();
                                            const proxyInfo = entry === 'DIRECT'
                                                ? undefined
                                                : { address: entry.address, port: entry.port };
                                            self.server.emit(EVENTS.REMOTE_SOCKET_ERROR, {
                                                err,
                                                destination: dest,
                                                args,
                                                originInfo,
                                                destinationInfo,
                                                proxyInfo,
                                            });
                                        },
                                    },
                                });
                                clearConnectionTimeout();
                                exitDomainOnce();
                                // Register lifecycle cleanup only for the successful
                                // destination (failed attempts' destinations were already
                                // destroyed by the inner functions).
                                if (successfulDestination && !successfulDestination.destroyed) {
                                    socket.on('close', () => {
                                        successfulDestination.destroy();
                                        const idx = self.destinationSockets.indexOf(successfulDestination);
                                        if (idx !== -1)
                                            self.destinationSockets.splice(idx, 1);
                                    });
                                }
                                return;
                            }
                            // chain was null/empty — fall through to existing dispatcher
                            // logic (SOCKS5 chain if configured, else DIRECT)
                        }
                        // === end CLIENT-SIDE HTTP CONNECT branch ===
                        let createExecuterAndRegisterToEvents = null;
                        let proxyServer = undefined;
                        // whitelisted addresses should never pass through a proxy
                        if (!isWhitelisted && ((_a = self.options.proxyServers) === null || _a === void 0 ? void 0 : _a.getAvailableProxy())) {
                            proxyServer = {
                                ...self.options.proxyServers.getAvailableProxy(),
                                shouldUseProxy: self.options.proxyServers.shouldUseProxy || _alwaysTrue,
                            };
                        }
                        const shouldUseProxy = (proxyServer === null || proxyServer === void 0 ? void 0 : proxyServer.shouldUseProxy({
                            socket,
                            destinationInfo,
                            originInfo,
                        })) || false;
                        const proxyInfo = proxyServer && shouldUseProxy
                            ? {
                                address: proxyServer.address,
                                port: proxyServer.port,
                            }
                            : undefined;
                        const basicEventData = {
                            originInfo,
                            destinationInfo,
                            proxyInfo,
                        };
                        self.server.emit(EVENTS.ACCEPT_NEW_REQUEST, basicEventData);
                        if (proxyServer && shouldUseProxy) {
                            const retries = { handshake: 0, rpcReply: 0 };
                            proxyServer.onHandshakeCompleted = () => {
                                self.server.emit(EVENTS.REMOTE_PROXY_HANDSHAKE_COMPLETED, basicEventData);
                            };
                            proxyServer.onHandshakeTimeout = () => {
                                self.server.emit(EVENTS.REMOTE_PROXY_HANDSHAKE_TIMEOUT, {
                                    ...basicEventData,
                                    timeout: proxyServer.handshakeTimeout,
                                });
                                retries.handshake++;
                                if (retries.handshake <= 3)
                                    createExecuterAndRegisterToEvents();
                                else
                                    end(constants_1.RFC_1928_REPLIES.NETWORK_UNREACHABLE, args);
                            };
                            proxyServer.onRpc1928ReplyTimeout = () => {
                                self.server.emit(EVENTS.REMOTE_PROXY_RFC_1928_REPLY_TIMEOUT, {
                                    ...basicEventData,
                                    timeout: proxyServer.rpcReplyTimeout,
                                });
                                retries.rpcReply++;
                                if (retries.rpcReply <= 3)
                                    createExecuterAndRegisterToEvents();
                                else
                                    end(constants_1.RFC_1928_REPLIES.NETWORK_UNREACHABLE, args);
                            };
                            proxyServer.onRpc1928Reply = (reply) => {
                                self.server.emit(EVENTS.REMOTE_PROXY_RFC_1928_REPLY, {
                                    ...basicEventData,
                                    reply,
                                });
                            };
                        }
                        createExecuterAndRegisterToEvents = () => {
                            const connectionTimeout = 120000;
                            let connectionTimeoutId = setTimeout(() => self.server.emit(EVENTS.REMOTE_CONNECTION_TIMEOUT_ERROR, {
                                ...basicEventData,
                                timeout: connectionTimeout,
                            }), connectionTimeout);
                            let destination;
                            const clearConnectionTimeout = () => {
                                connectionTimeoutId && clearTimeout(connectionTimeoutId);
                                connectionTimeoutId = null;
                            };
                            const onConnect = (destinationSocket) => {
                                connectionFilterDomain.exit();
                                clearConnectionTimeout();
                                // close destination socket when after original socket was closed
                                socket.on('close', () => {
                                    destinationSocket.destroy();
                                    self.destinationSockets.splice(self.destinationSockets.indexOf(destinationSocket), 1);
                                });
                                const event = shouldUseProxy
                                    ? EVENTS.CONNECTED_TO_REMOTE_PROXY
                                    : EVENTS.CONNECTED_TO_REMOTE_ADDRESS;
                                self.server.emit(event, { destinationInfo, originInfo, proxyInfo });
                            };
                            const onError = (err) => {
                                // exit the connection filter domain
                                connectionFilterDomain.exit();
                                self.server.emit(EVENTS.REMOTE_SOCKET_ERROR, {
                                    err,
                                    destination,
                                    args,
                                    originInfo,
                                    destinationInfo,
                                    proxyInfo,
                                });
                            };
                            if (!shouldUseProxy) {
                                destination = (0, execute_request_1.executeRequest)({
                                    originalSocket: socket,
                                    addressType: args.atyp,
                                    address: args.dst.addr,
                                    port: args.dst.port,
                                    requestBuffer: args.requestBuffer,
                                    end,
                                    onConnect,
                                    onError,
                                });
                            }
                            else {
                                destination = (0, execute_request_through_remote_proxy_1.executeRequestThroughRemoteProxy)({
                                    originalSocket: socket,
                                    clientHandshakeBuffer,
                                    proxyServer: proxyServer,
                                    addressType: args.atyp,
                                    address: args.dst.addr,
                                    port: args.dst.port,
                                    requestBuffer: args.requestBuffer,
                                    end,
                                    onConnect,
                                    onError,
                                });
                            }
                            self.destinationSockets.push(destination);
                        };
                        createExecuterAndRegisterToEvents();
                    }));
                };
                (0, parse_connect_buffer_1.parseConnectBuffer)({
                    buffer,
                    onReqesutBufferParsed: () => self.activeSessions.push(socket),
                    onParseCompleted,
                    end,
                });
            }
            /**
             * +----+-----+-------+------+----------+----------+
             * |VER | REP |  RSV  | ATYP | BND.ADDR | BND.PORT |
             * +----+-----+-------+------+----------+----------+
             * | 1  |  1  | X'00' |  1   | Variable |    2     |
             * +----+-----+-------+------+----------+----------+
             *
             * @param {Buffer} response - a buffer representing the response
             * @param {object} args - arguments to supply to the proxy end event
             * @returns {undefined}
             **/
            function end(response, args) {
                (0, end_socket_connection_1.endConnection)({ socket, response, args });
                // indicate end of connection
                self.server.emit(EVENTS.PROXY_END, response, args);
            }
            /**
             * +----+----------+----------+
             * |VER | NMETHODS | METHODS  |
             * +----+----------+----------+
             * | 1  |    1     | 1 to 255 |
             * +----+----------+----------+
             *
             * @param {Buffer} buffer - a buffer
             * @returns {undefined}
             **/
            function _handshake(buffer) {
                clientHandshakeBuffer = buffer;
                const onHandshakeCompleted = ({ clientHandshakeBuffer, responseBuffer, }) => {
                    self.server.emit(EVENTS.HANDSHAKE, { socket, clientHandshakeBuffer, responseBuffer });
                };
                (0, handshake_1.handshake)({
                    socket,
                    buffer,
                    options: self.options,
                    connect,
                    authenticate: _authenticate,
                    onHandshakeCompleted,
                    end,
                });
            }
            socket.once('data', _handshake);
            // capture socket closure
            socket.once('end', () => {
                // remove the session from currently the active sessions list
                self.activeSessions.splice(self.activeSessions.indexOf(socket), 1);
            });
        });
    }
    set shouldBlockPrivateAddresses(value) {
        this.options.shouldBlockPrivateAddresses = value;
    }
    get shouldBlockPrivateAddresses() {
        return !!this.options.shouldBlockPrivateAddresses;
    }
    setShouldWhitelist(shouldWhitelist) {
        this.options.shouldWhitelist = shouldWhitelist;
    }
    setShouldBlockRequest(shouldBlockRequest) {
        this.options.shouldBlockRequest = shouldBlockRequest;
    }
    get server() {
        return this._server;
    }
}
exports.SocksServer = SocksServer;
function createSocks5ProxyServer(options = {}) {
    const socksServer = new SocksServer(options);
    return socksServer;
}
exports.EVENTS = __importStar(require("./events"));
