// @applitools/eg-socks5-proxy-server/lib/try-http-proxy-chain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tryHttpProxyChain = tryHttpProxyChain;
exports.normalizeChain = normalizeChain;
const constants_1 = require("./constants");
const execute_request_1 = require("./execute-request");
const execute_request_through_http_proxy_1 = require("./execute-request-through-http-proxy");
// RFC 1928 reply codes we consider RETRYABLE — i.e. the chain should move to
// the next entry. These correspond to transport/proxy availability problems.
// Anything else (notably 407 → CONNECTION_NOT_ALLOWED) is TERMINAL: it tells
// us "this specific proxy rejected the request for a reason that won't be
// fixed by trying a different proxy."
const RETRYABLE_RFC_CODES = new Set([
    constants_1.RFC_1928_REPLIES.NETWORK_UNREACHABLE, // 502/503 or convertSocketErrorToRpcReply default
    constants_1.RFC_1928_REPLIES.HOST_UNREACHABLE, // EADDRNOTAVAIL
    constants_1.RFC_1928_REPLIES.CONNECTION_REFUSED, // ECONNREFUSED or proxy hang-up
    constants_1.RFC_1928_REPLIES.TTL_EXPIRED, // 504, header timeout
    constants_1.RFC_1928_REPLIES.GENERAL_FAILURE, // malformed status line, header overflow
]);
/**
 * Iterates the chain. Returns the destination socket of the successful
 * attempt (for the dispatcher to attach lifecycle handlers), or
 * `undefined` if the chain exhausted without success.
 *
 * Side effects (which the dispatcher relies on):
 *   - On success: `executeRequestThroughHttpProxy` (or `executeRequest`)
 *     has already written the SOCKS5 SUCCEEDED reply to `originalSocket`
 *     and wired the bidirectional pipe.
 *   - On terminal/exhausted failure: `params.end(rfc, ...)` has been
 *     called with the last RFC 1928 reply code.
 *   - For every attempt: callbacks fire as documented.
 */
async function tryHttpProxyChain(params) {
    let lastRfcReply = constants_1.RFC_1928_REPLIES.GENERAL_FAILURE;
    let lastDestination;
    for (let i = 0; i < params.chain.length; i++) {
        const entry = params.chain[i];
        if (entry === 'DIRECT') {
            const outcome = await runDirectEntry(params);
            lastRfcReply = outcome.rfc;
            lastDestination = outcome.destination;
            if (outcome.kind === 'success')
                return outcome.destination;
            // DIRECT failure is terminal — there's nothing to fall back to.
            params.end(outcome.rfc, { requestBuffer: params.requestBuffer });
            return outcome.destination;
        }
        const outcome = await runProxyEntry(entry, params);
        lastRfcReply = outcome.rfc;
        lastDestination = outcome.destination;
        if (outcome.kind === 'success')
            return outcome.destination;
        if (outcome.kind === 'terminal') {
            params.end(outcome.rfc, { requestBuffer: params.requestBuffer });
            return outcome.destination;
        }
        // retryable — try next entry
    }
    // Chain exhausted with no success — surface the last failure.
    params.end(lastRfcReply, { requestBuffer: params.requestBuffer });
    return lastDestination;
}
function runProxyEntry(entry, params) {
    return new Promise((resolve) => {
        var _a;
        let settled = false;
        // We capture the destination synchronously below — but the closure here
        // needs forward-declared visibility so the callbacks can reference it.
        let destinationRef;
        const settle = (outcome) => {
            if (settled)
                return;
            settled = true;
            resolve(outcome);
        };
        const destination = (0, execute_request_through_http_proxy_1.executeRequestThroughHttpProxy)({
            originalSocket: params.originalSocket,
            proxy: entry,
            addressType: params.addressType,
            address: params.address,
            port: params.port,
            requestBuffer: params.requestBuffer,
            onConnect: (sock) => {
                var _a, _b;
                // TCP/TLS to the proxy is up; surface the lifecycle event so the
                // dispatcher can clear timeouts and register cleanup.
                (_b = (_a = params.callbacks).onConnectedToHttpProxy) === null || _b === void 0 ? void 0 : _b.call(_a, entry, sock);
            },
            onProxyResponded: (statusCode) => {
                var _a, _b;
                (_b = (_a = params.callbacks).onProxyResponded) === null || _b === void 0 ? void 0 : _b.call(_a, statusCode, entry);
                if (statusCode >= 200 && statusCode < 300) {
                    // The inner function will write the SOCKS5 SUCCEEDED frame and
                    // wire the pipe in the same synchronous block. Resolve as
                    // success — the pipe will be live by the next tick.
                    settle({ kind: 'success', destination: destinationRef, rfc: statusCode });
                }
            },
            onProxyFailed: (rfc) => {
                var _a, _b;
                (_b = (_a = params.callbacks).onProxyFailed) === null || _b === void 0 ? void 0 : _b.call(_a, rfc, entry);
                const isRetryable = RETRYABLE_RFC_CODES.has(rfc);
                settle({
                    kind: isRetryable ? 'retryable' : 'terminal',
                    destination: destinationRef,
                    rfc,
                });
            },
            // NO-OP `end` while iterating. The orchestrator will call the real
            // `params.end` once chain-level outcome is decided.
            end: () => {
                /* deliberately empty — see file-level comment */
            },
            onError: (err) => {
                var _a, _b;
                (_b = (_a = params.callbacks).onSocketError) === null || _b === void 0 ? void 0 : _b.call(_a, err, entry, destinationRef);
            },
        });
        destinationRef = destination;
        (_a = params.registerDestination) === null || _a === void 0 ? void 0 : _a.call(params, destination);
    });
}
function runDirectEntry(params) {
    return new Promise((resolve) => {
        var _a;
        let settled = false;
        let destinationRef;
        const settle = (outcome) => {
            if (settled)
                return;
            settled = true;
            resolve(outcome);
        };
        const destination = (0, execute_request_1.executeRequest)({
            originalSocket: params.originalSocket,
            addressType: params.addressType,
            address: params.address,
            port: params.port,
            requestBuffer: params.requestBuffer,
            end: (rfc) => {
                // DIRECT path failed before establishing — surface as terminal.
                settle({ kind: 'terminal', destination: destinationRef, rfc });
            },
            onConnect: (sock) => {
                var _a, _b;
                // `executeRequest` writes the SOCKS5 SUCCEEDED reply and sets up
                // the pipe BEFORE invoking onConnect. So we can settle as success
                // right here.
                (_b = (_a = params.callbacks).onConnectedToRemoteAddress) === null || _b === void 0 ? void 0 : _b.call(_a, sock);
                settle({
                    kind: 'success',
                    destination: destinationRef,
                    rfc: constants_1.RFC_1928_REPLIES.SUCCEEDED,
                });
            },
            onError: (err) => {
                var _a, _b;
                (_b = (_a = params.callbacks).onSocketError) === null || _b === void 0 ? void 0 : _b.call(_a, err, 'DIRECT', destinationRef);
            },
        });
        destinationRef = destination;
        (_a = params.registerDestination) === null || _a === void 0 ? void 0 : _a.call(params, destination);
    });
}
/**
 * Normalize a resolver result into a chain.
 *   - `null` / `undefined` / `[]` → `null` (caller falls through)
 *   - single decision → `[decision]`
 *   - array → used as-is
 */
function normalizeChain(result) {
    if (result == null)
        return null;
    if (Array.isArray(result))
        return result.length === 0 ? null : result;
    return [result];
}
