// @applitools/core/dist/ufg/utils/media-playback-guard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pagePayload = exports.makeMediaPlaybackGuard = void 0;
/**
 * Prevents media playback side-effects during the layoutBreakpoints snapshot loop.
 *
 * dom-snapshot calls video.play() per breakpoint to capture a video frame. For
 * BrightCove / video.js players this triggers CSS class changes (has-started, is-playing)
 * and subtitle text injection that outlast the corresponding pause() call. With multiple
 * breakpoints the mutations accumulate and corrupt visual snapshots.
 *
 * Lifecycle:
 *   const guard = makeMediaPlaybackGuard({context, logger})
 *   await guard.setup()                          // once, before the loop
 *   for (const entry of entries) {
 *     await guard.beforeResize()                 // before setViewportSize
 *     // setViewportSize / reload / lazyLoad
 *     await guard.beforeSnapshot()               // after resize, before takeDomSnapshot
 *     // takeDomSnapshot
 *     await guard.afterSnapshot()                // after takeDomSnapshot
 *   }
 *   // setViewportSize (restore)
 *   await guard.teardown()                       // once, after final viewport restore
 *
 * Implementation notes:
 *
 * The TypeScript layer is just orchestration — `setup` / `beforeResize` / etc. each
 * dispatch a phase name to a single self-contained page-side payload (`pagePayload`
 * below) via `context.execute`. The first call installs a library at `window[ns]`
 * with one handler per lifecycle method; subsequent calls just look up and run
 * `window[ns][phase]()`. Teardown deletes `window[ns]` so a fresh guard starts clean.
 *
 *   - `ns` is randomised per guard instance so concurrent guards on the same page
 *     (e.g. nested checks) do not collide.
 *   - Per-element state lives in a `WeakMap`, not on the elements themselves — no
 *     marker properties leak onto the page's DOM.
 *   - A single `walkAll(root, fn)` helper handles every Document/ShadowRoot traversal.
 *   - A `MutationObserver` covers the gap between phase calls (e.g. videos that
 *     BrightCove creates during setViewportSize). The synchronous walks in
 *     `beforeSnapshot` remain because the observer fires on a microtask and would
 *     otherwise lose the race against dom-snapshot's serialization.
 */
function makeMediaPlaybackGuard({ context, logger, }) {
    const ns = '__applitools_mpg_' + Math.random().toString(36).slice(2) + '__';
    return {
        setup: () => runPhase('setup'),
        beforeResize: () => runPhase('beforeResize'),
        beforeSnapshot: () => runPhase('beforeSnapshot'),
        afterSnapshot: () => runPhase('afterSnapshot'),
        teardown: () => runPhase('teardown'),
    };
    function runPhase(phase) {
        return context
            .execute(pagePayload, { ns, phase })
            .catch((err) => logger.warn(`media guard: ${phase} failed`, err));
    }
}
exports.makeMediaPlaybackGuard = makeMediaPlaybackGuard;
/**
 * The entire page-side library — a single self-contained function shipped to the
 * browser by `context.execute`. No closure variables from the TypeScript side are
 * visible inside this body.
 *
 * Contract: on first invocation, installs `window[ns]` with `{setup, beforeResize,
 * beforeSnapshot, afterSnapshot, teardown}` handlers; on every invocation, calls
 * `window[ns][phase]()` if present. The handlers all close over the same set of
 * private helpers and a `WeakMap`-backed per-element state.
 */
/* istanbul ignore next — runs only inside the browser, not in node */
function pagePayload(arg) {
    const w = window;
    const ns = arg.ns;
    if (!w[ns])
        install();
    const state = w[ns];
    if (state && typeof state[arg.phase] === 'function')
        state[arg.phase]();
    function install() {
        // ----- private state -----
        const videoUndos = new WeakMap();
        const captionUndos = new WeakMap();
        const playerSnapshots = [];
        const originalProtoPlay = HTMLMediaElement.prototype.play;
        const originalCreateElement = document.createElement.bind(document);
        const originalCreateElementNS = document.createElementNS.bind(document);
        let observer = null;
        // ----- helpers -----
        /** Single DOM traversal helper. Visits every element under `root`, recursing into shadow roots. */
        function walkAll(root, fn) {
            const els = root.querySelectorAll('*');
            for (let i = 0; i < els.length; i++) {
                const el = els[i];
                fn(el);
                const sr = el.shadowRoot;
                if (sr)
                    walkAll(sr, fn);
            }
        }
        function walkVideos(root, fn) {
            walkAll(root, el => {
                if (el.tagName === 'VIDEO')
                    fn(el);
            });
        }
        function walkCaptions(root, fn) {
            walkAll(root, el => {
                const cl = el.classList;
                if (cl && cl.contains('vjs-text-track-display'))
                    fn(el);
            });
        }
        /**
         * Replace an own-property on `obj` with `replacement` and return a closure that
         * puts the original back. Restores correctly whether the original lived on the
         * instance or only on the prototype.
         */
        function overrideOwn(obj, key, replacement) {
            const wasOwn = Object.prototype.hasOwnProperty.call(obj, key);
            const original = wasOwn ? obj[key] : undefined;
            obj[key] = replacement;
            return function restore() {
                if (wasOwn)
                    obj[key] = original;
                else
                    delete obj[key];
            };
        }
        function runUndos(undos) {
            for (let i = undos.length - 1; i >= 0; i--) {
                try {
                    undos[i]();
                }
                catch (_) {
                    // best-effort restore: a failure here must not block the rest
                }
            }
        }
        // ----- locking primitives -----
        function lockVideo(video) {
            if (videoUndos.has(video))
                return;
            const undos = [];
            // 1. Mask the `autoplay` IDL attribute so reads return false and writes are dropped.
            const hadAutoplay = video.hasAttribute('autoplay');
            if (hadAutoplay)
                video.removeAttribute('autoplay');
            try {
                Object.defineProperty(video, 'autoplay', {
                    configurable: true,
                    get: () => false,
                    set: () => { },
                });
            }
            catch (_) { }
            undos.push(() => {
                try {
                    delete video.autoplay;
                }
                catch (_) { }
                if (hadAutoplay)
                    video.setAttribute('autoplay', '');
            });
            // 2. setAttribute('autoplay', ...) — BrightCove uses this directly, bypassing the IDL setter.
            undos.push(overrideOwn(video, 'setAttribute', function (name, value) {
                if (name === 'autoplay')
                    return;
                return Element.prototype.setAttribute.call(this, name, value);
            }));
            // 3. Make captureStream() return zero video tracks so dom-snapshot's buildBlobVideos
            //    skips its play()/record/pause cycle entirely.
            if (video.captureStream) {
                undos.push(overrideOwn(video, 'captureStream', function () {
                    return { getVideoTracks: () => [] };
                }));
            }
            // 4. Disable text tracks (caption rendering source).
            const tracks = video.textTracks;
            const savedModes = [];
            for (let i = 0; i < tracks.length; i++) {
                savedModes.push(tracks[i].mode);
                tracks[i].mode = 'disabled';
            }
            undos.push(() => {
                const t = video.textTracks;
                for (let i = 0; i < t.length && i < savedModes.length; i++)
                    t[i].mode = savedModes[i];
            });
            videoUndos.set(video, undos);
        }
        function unlockVideo(video) {
            const undos = videoUndos.get(video);
            if (!undos)
                return;
            videoUndos.delete(video);
            runUndos(undos);
        }
        /**
         * Lock a .vjs-text-track-display caption container. Instance-level overrides because
         * BrightCove's event handlers re-inject cue elements in microtasks between our
         * phase call returning and dom-snapshot serializing the DOM — synchronous, persistent
         * overrides intercept every insertion path regardless of timing.
         */
        function lockCaption(c) {
            if (captionUndos.has(c))
                return;
            const undos = [];
            while (c.firstChild)
                c.removeChild(c.firstChild);
            undos.push(overrideOwn(c, 'appendChild', (n) => n));
            undos.push(overrideOwn(c, 'insertBefore', (n) => n));
            undos.push(overrideOwn(c, 'insertAdjacentHTML', () => { }));
            undos.push(overrideOwn(c, 'insertAdjacentElement', (_pos, n) => n));
            try {
                Object.defineProperty(c, 'innerHTML', {
                    configurable: true,
                    get: () => '',
                    set: () => { },
                });
                undos.push(() => {
                    try {
                        delete c.innerHTML;
                    }
                    catch (_) { }
                });
            }
            catch (_) { }
            captionUndos.set(c, undos);
        }
        function unlockCaption(c) {
            const undos = captionUndos.get(c);
            if (!undos)
                return;
            captionUndos.delete(c);
            runUndos(undos);
        }
        // ----- player container snapshot / restore -----
        function snapshotPlayers() {
            walkVideos(document, video => {
                const playerEl = video.closest('video-js, [data-vjs-player], .video-js');
                playerSnapshots.push({
                    video,
                    paused: video.paused,
                    currentTime: video.currentTime,
                    playerEl,
                    playerClasses: playerEl ? Array.from(playerEl.classList) : [],
                });
            });
        }
        function restorePlayers() {
            for (let i = 0; i < playerSnapshots.length; i++) {
                const s = playerSnapshots[i];
                if (s.paused && !s.video.paused) {
                    s.video.pause();
                    try {
                        s.video.currentTime = s.currentTime;
                    }
                    catch (_) { } // MSE may reject seeks
                }
                if (s.playerEl) {
                    const current = Array.from(s.playerEl.classList);
                    for (let j = 0; j < current.length; j++) {
                        if (s.playerClasses.indexOf(current[j]) === -1)
                            s.playerEl.classList.remove(current[j]);
                    }
                    for (let j = 0; j < s.playerClasses.length; j++)
                        s.playerEl.classList.add(s.playerClasses[j]);
                }
            }
        }
        // ----- global patches (live for the entire loop) -----
        function patchGlobals() {
            // video.js calls HTMLMediaElement.prototype.play.call(el) directly — prototype patch is mandatory.
            // Stays alive through teardown's restorePlayers() so BrightCove's pause-event handler
            // cannot call the (now-live) native play() and restart playback.
            ;
            HTMLMediaElement.prototype.play = function () {
                return Promise.resolve();
            };
            document.createElement = function (tag) {
                const el = originalCreateElement(tag);
                if (typeof tag === 'string' && tag.toLowerCase() === 'video') {
                    lockVideo(el);
                }
                return el;
            };
            document.createElementNS = function (nsArg, tag) {
                const el = originalCreateElementNS(nsArg, tag);
                if (typeof tag === 'string' && tag.toLowerCase() === 'video') {
                    lockVideo(el);
                }
                return el;
            };
        }
        function unpatchGlobals() {
            ;
            HTMLMediaElement.prototype.play = originalProtoPlay;
            document.createElement = originalCreateElement;
            document.createElementNS = originalCreateElementNS;
        }
        // ----- mutation observer (belt-and-suspenders for late-arriving nodes) -----
        function startObserver() {
            observer = new MutationObserver(mutations => {
                for (let i = 0; i < mutations.length; i++) {
                    const added = mutations[i].addedNodes;
                    for (let j = 0; j < added.length; j++) {
                        const node = added[j];
                        if (!(node instanceof Element))
                            continue;
                        if (node.tagName === 'VIDEO')
                            lockVideo(node);
                        else if (node.classList && node.classList.contains('vjs-text-track-display'))
                            lockCaption(node);
                        walkVideos(node, lockVideo);
                        walkCaptions(node, lockCaption);
                    }
                }
            });
            observer.observe(document, { childList: true, subtree: true });
        }
        function stopObserver() {
            if (observer) {
                observer.disconnect();
                observer = null;
            }
        }
        // ----- bootstrap -----
        // Player state is NOT snapshotted here. It is captured fresh in `beforeSnapshot`
        // so that responsive classes (vjs-layout-small, vjs-fluid, orientation classes)
        // applied during setViewportSize are preserved across afterSnapshot's restore.
        // This also automatically covers videos BrightCove injects mid-loop, since the
        // walk re-runs each iteration.
        patchGlobals();
        walkCaptions(document, lockCaption);
        startObserver();
        // ----- phase handlers (called from TS via runPhase) -----
        w[ns] = {
            // `setup` is already complete by the time we get here; the no-op handler keeps the dispatch shape uniform.
            setup() { },
            beforeResize() {
                walkVideos(document, lockVideo);
                walkCaptions(document, lockCaption);
            },
            beforeSnapshot() {
                // Safety net first — pause videos and disable tracks before snapshotting state,
                // so playback classes (vjs-playing) settle to a consistent paused-state.
                walkVideos(document, video => {
                    if (!video.paused)
                        video.pause();
                    const t = video.textTracks;
                    for (let i = 0; i < t.length; i++)
                        t[i].mode = 'disabled';
                });
                // Capture fresh per-player state — runs after resize so responsive classes
                // reflect the current breakpoint, and so videos BrightCove injected during
                // the resize are included.
                playerSnapshots.length = 0;
                snapshotPlayers();
                // MutationObserver is async; lock new caption containers synchronously before
                // dom-snapshot's next-macrotask serialization runs.
                walkCaptions(document, lockCaption);
            },
            afterSnapshot() {
                walkVideos(document, unlockVideo);
                restorePlayers();
                // Clear after consume — the captured state is breakpoint-specific and must not
                // leak into the next iteration or teardown, where the viewport differs.
                playerSnapshots.length = 0;
            },
            teardown() {
                // Restore players BEFORE unpatching globals: BrightCove's pause-event handler can call
                // play() during pause(), and we need that call to still hit the no-op prototype patch.
                restorePlayers();
                stopObserver();
                unpatchGlobals();
                walkVideos(document, unlockVideo);
                walkCaptions(document, unlockCaption);
                delete w[ns];
            },
        };
    }
}
exports.pagePayload = pagePayload;
