// @pact-foundation/pact/src/v3/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTPResponseStatusClass = exports.SpecificationVersion = void 0;
var SpecificationVersion;
(function (SpecificationVersion) {
    SpecificationVersion[SpecificationVersion["SPECIFICATION_VERSION_V2"] = 3] = "SPECIFICATION_VERSION_V2";
    SpecificationVersion[SpecificationVersion["SPECIFICATION_VERSION_V3"] = 4] = "SPECIFICATION_VERSION_V3";
    SpecificationVersion[SpecificationVersion["SPECIFICATION_VERSION_V4"] = 5] = "SPECIFICATION_VERSION_V4";
})(SpecificationVersion || (exports.SpecificationVersion = SpecificationVersion = {}));
/**
 * Enum for HTTP response status classes used with the status code matcher.
 * These values correspond to the status code ranges defined in RFC 7231.
 */
var HTTPResponseStatusClass;
(function (HTTPResponseStatusClass) {
    // Informational responses (100–199)
    HTTPResponseStatusClass["Information"] = "information";
    // Successful responses (200–299)
    HTTPResponseStatusClass["Success"] = "success";
    // Redirects (300–399)
    HTTPResponseStatusClass["Redirect"] = "redirect";
    // Client errors (400–499)
    HTTPResponseStatusClass["ClientError"] = "clientError";
    // Server errors (500–599)
    HTTPResponseStatusClass["ServerError"] = "serverError";
    // Non-error response (< 400)
    HTTPResponseStatusClass["NonError"] = "nonError";
    // Any error response (>= 400)
    HTTPResponseStatusClass["Error"] = "error";
})(HTTPResponseStatusClass || (exports.HTTPResponseStatusClass = HTTPResponseStatusClass = {}));
