// @applitools/eg-socks5-proxy-server/src/execute-request-through-http-proxy.ts
/**
 * HTTP CONNECT upstream execution path for the SOCKS5 server.
 *
 * CLIENT-SIDE: used when a `SocksServer` is deployed inside a tunnel agent
 * (e.g. `@applitools/execution-grid-tunnel`) on a customer machine, and the
 * target AUT is reachable only through a corporate HTTP CONNECT proxy.
 *
 * This file is INTENTIONALLY separate from `execute-request-through-remote-proxy.ts`
 * which speaks the SOCKS5 sub-protocol. That one is used SERVER-SIDE inside the EC
 * browser pod (`@applitools/eg-chromedriver-image`) to load-balance the session's
 * traffic across multiple tunnels — a fundamentally different topology.
 *
 * Wire protocol against the upstream proxy (after the transport — plain TCP or
 * TLS, depending on `proxy.tls` — is up):
 *
 *     CONNECT host:port HTTP/1.1
 *     Host: host:port
 *     [Proxy-Authorization: Basic <base64(user:pass)>]
 *     Proxy-Connection: keep-alive
 *
 *     HTTP/1.x 200 Connection established
 *     ...
 *
 * After the 200 response, bytes flow opaquely both directions. TLS handshakes
 * for HTTPS targets occur END-TO-END between the originating SOCKS5 client
 * (typically Chrome in the EC browser pod) and the destination server — the
 * corporate proxy never sees plaintext.
 *
 * If `proxy.tls` is true (PAC's `HTTPS host:port` directive), the entire
 * exchange above is wrapped in a TLS session between this process and the
 * proxy — the `Proxy-Authorization` header is encrypted in transit. The
 * inner end-to-end TLS for HTTPS targets is still preserved.
 *
 * On a non-2xx status, the proxy's HTTP code is mapped to an RFC 1928 reply
 * which is sent to the originating SOCKS5 client so existing SOCKS5 error
 * semantics propagate cleanly through the tunnel.
 */
import net, {Socket} from 'net'
import tls from 'tls'

import {RFC_1928_REPLIES} from './constants'
import {convertSocketErrorToRpcReply} from './utils/convert-socket-error-to-rpc-reply'

import {ExecuteReqeustError} from '../types/execute-request'
import {
  ExecuteRequestThroughHttpProxyParameters,
  HttpProxyDecision,
} from '../types/execute-request-through-http-proxy'

const HEADER_TERMINATOR = Buffer.from('\r\n\r\n')
const STATUS_LINE_REGEX = /^HTTP\/1\.[01]\s+(\d{3})\s/i
const DEFAULT_HEADER_TIMEOUT_MS = 30000
// Defensive cap. Corporate proxies sometimes echo many response headers; this is
// the byte limit before we see the \r\n\r\n terminator. 16 KiB is generous.
const MAX_HEADER_BYTES = 16 * 1024

export function executeRequestThroughHttpProxy({
  originalSocket,
  proxy,
  addressType,
  address,
  port,
  requestBuffer,
  onProxyResponded,
  onProxyFailed,
  end,
  onError,
  onConnect,
}: ExecuteRequestThroughHttpProxyParameters): Socket {
  let headerTimeoutId: NodeJS.Timeout | null = null
  let headerBuf = Buffer.alloc(0)
  // Becomes true once we've seen `HTTP/1.x 2xx` and written the SOCKS5 success
  // frame back to the originating client. After this point, the destination
  // socket is in `pipe()` mode and we do not treat further inbound data as
  // proxy-protocol bytes.
  let tunnelEstablished = false
  // Set once we've reported a failure (either via the `error` event branch or
  // via the response-parser branches below). Used to avoid double-reporting if
  // both `error` and `close` fire after `destroy()`.
  let failureReported = false

  // Send the CONNECT request once the transport (raw TCP, or post-handshake
  // TLS) is ready. For plain proxies this fires on `connect`; for TLS proxies
  // it fires on `secureConnect` — the difference is hidden behind the
  // factory's connectListener.
  const onTransportReady = () => {
    destination.write(buildConnectRequest(address, port, proxy))
    headerTimeoutId = setTimeout(onHeaderTimeout, DEFAULT_HEADER_TIMEOUT_MS)
    destination.on('data', onProxyData)
    onConnect(destination)
  }

  // Open the transport. For PAC's `HTTPS host:port` (proxy.tls === true) we
  // wrap the connection in TLS so the CONNECT exchange itself is encrypted —
  // the corporate proxy is verified against the host's default trust
  // anchors. Otherwise plain TCP (PAC's `PROXY host:port`).
  const destination: Socket = proxy.tls
    ? tls.connect(
        {host: proxy.address, port: proxy.port, ALPNProtocols: ['http/1.1']},
        onTransportReady,
      )
    : net.createConnection(proxy.port, proxy.address, onTransportReady)

  destination.setKeepAlive(true)

  destination.on('error', (err: ExecuteReqeustError) => {
    if (headerTimeoutId) {
      clearTimeout(headerTimeoutId)
      headerTimeoutId = null
    }
    if (tunnelEstablished) {
      // The pipe is already in place. Just surface the error; don't translate
      // to a SOCKS5 reply (one was already sent).
      onError(annotateError(err))
      return
    }
    if (failureReported) return
    failureReported = true
    onError(annotateError(err))
    const rfcReply = convertSocketErrorToRpcReply(err)
    onProxyFailed(rfcReply)
    end(rfcReply, {requestBuffer})
  })

  // If the proxy closes the socket gracefully (e.g. FIN with no body) BEFORE
  // we've completed the CONNECT response, treat it as a connection failure.
  // Without this branch, a half-open or silently-dropped proxy would leave the
  // originating SOCKS5 client hanging.
  destination.on('close', () => {
    if (headerTimeoutId) {
      clearTimeout(headerTimeoutId)
      headerTimeoutId = null
    }
    if (tunnelEstablished || failureReported) return
    failureReported = true
    onProxyFailed(RFC_1928_REPLIES.CONNECTION_REFUSED)
    end(RFC_1928_REPLIES.CONNECTION_REFUSED, {requestBuffer})
  })

  return destination

  // ------------------------------------------------------------------
  // helpers, closed over the per-request state above

  function annotateError(err: ExecuteReqeustError): ExecuteReqeustError {
    err.addr = address
    err.atyp = addressType
    err.port = port
    return err
  }

  function onHeaderTimeout(): void {
    headerTimeoutId = null
    if (tunnelEstablished || failureReported) return
    failureReported = true
    onProxyFailed(RFC_1928_REPLIES.TTL_EXPIRED)
    end(RFC_1928_REPLIES.TTL_EXPIRED, {requestBuffer})
    destination.destroy()
  }

  function onProxyData(chunk: Buffer): void {
    if (tunnelEstablished) return // shouldn't fire; listener was removed below

    headerBuf = Buffer.concat([headerBuf, chunk])
    const terminatorIdx = headerBuf.indexOf(HEADER_TERMINATOR)
    if (terminatorIdx === -1) {
      // Header not complete yet. Guard against unbounded buffering — a
      // misbehaving proxy that never sends \r\n\r\n must not OOM the process.
      if (headerBuf.length > MAX_HEADER_BYTES) {
        if (headerTimeoutId) clearTimeout(headerTimeoutId)
        headerTimeoutId = null
        if (!failureReported) {
          failureReported = true
          const rfc = RFC_1928_REPLIES.GENERAL_FAILURE
          onProxyFailed(rfc)
          end(rfc, {requestBuffer})
        }
        destination.destroy()
      }
      return
    }

    if (headerTimeoutId) clearTimeout(headerTimeoutId)
    headerTimeoutId = null
    destination.removeListener('data', onProxyData)

    const headerSection = headerBuf.slice(0, terminatorIdx).toString('latin1')
    const leftover = headerBuf.slice(terminatorIdx + HEADER_TERMINATOR.length)

    const statusLine = headerSection.split('\r\n', 1)[0]
    const statusMatch = STATUS_LINE_REGEX.exec(statusLine)
    if (!statusMatch) {
      // Couldn't parse a status line. Treat as proxy protocol failure.
      onProxyResponded(0)
      if (!failureReported) {
        failureReported = true
        const rfc = RFC_1928_REPLIES.GENERAL_FAILURE
        onProxyFailed(rfc)
        end(rfc, {requestBuffer})
      }
      destination.destroy()
      return
    }

    const statusCode = Number(statusMatch[1])
    onProxyResponded(statusCode)

    if (statusCode >= 200 && statusCode < 300) {
      // Tunnel established. Send SOCKS5 success to the originating client,
      // then any leftover header-trailing bytes (which already belong to the
      // tunneled stream), then wire the bidirectional pipe.
      tunnelEstablished = true
      const responseBuffer = Buffer.alloc(requestBuffer.length)
      requestBuffer.copy(responseBuffer)
      responseBuffer[1] = RFC_1928_REPLIES.SUCCEEDED
      originalSocket.write(responseBuffer, () => {
        if (leftover.length > 0) originalSocket.write(leftover)
        destination.pipe(originalSocket)
        originalSocket.pipe(destination)
      })
      return
    }

    // Non-2xx — map to the closest RFC 1928 reply code and fail the SOCKS5
    // request. Header may include `Proxy-Authenticate` etc., but we don't
    // surface those; the resolver / caller decided to use this proxy.
    const rfc = mapHttpStatusToRfc1928Reply(statusCode)
    if (!failureReported) {
      failureReported = true
      onProxyFailed(rfc)
      end(rfc, {requestBuffer})
    }
    destination.destroy()
  }
}

function buildConnectRequest(host: string, port: number, proxy: HttpProxyDecision): Buffer {
  const lines = [`CONNECT ${host}:${port} HTTP/1.1`, `Host: ${host}:${port}`]
  const authValue = formatProxyAuthorization(proxy.auth)
  if (authValue) lines.push(`Proxy-Authorization: ${authValue}`)
  lines.push('Proxy-Connection: keep-alive')
  lines.push('', '') // produces the trailing \r\n\r\n
  return Buffer.from(lines.join('\r\n'), 'latin1')
}

function formatProxyAuthorization(auth: HttpProxyDecision['auth']): string | undefined {
  if (!auth) return undefined
  if (typeof auth === 'string') {
    return `Basic ${Buffer.from(auth, 'utf8').toString('base64')}`
  }
  if (typeof auth === 'object' && auth.username !== undefined) {
    const creds = `${auth.username}:${auth.password ?? ''}`
    return `Basic ${Buffer.from(creds, 'utf8').toString('base64')}`
  }
  return undefined
}

function mapHttpStatusToRfc1928Reply(statusCode: number): number {
  if (statusCode === 407) return RFC_1928_REPLIES.CONNECTION_NOT_ALLOWED
  if (statusCode === 502) return RFC_1928_REPLIES.NETWORK_UNREACHABLE
  if (statusCode === 503) return RFC_1928_REPLIES.NETWORK_UNREACHABLE
  if (statusCode === 504) return RFC_1928_REPLIES.TTL_EXPIRED
  return RFC_1928_REPLIES.GENERAL_FAILURE
}
