// @applitools/eg-socks5-proxy-server/src/index.ts
import {
  RFC_1928_ATYP,
  RFC_1928_COMMANDS,
  RFC_1928_METHODS,
  RFC_1928_REPLIES,
  RFC_1928_VERSION,
  RFC_1929_REPLIES,
  RFC_1929_VERSION,
} from './constants'

import * as EVENTS from './events'
import {endConnection} from './end-socket-connection'
import {handshake} from './handshake'
import {authenticate} from './authenticate'
import {parseConnectBuffer} from './parse-connect-buffer'
import {executeRequest} from './execute-request'
import {executeRequestThroughRemoteProxy} from './execute-request-through-remote-proxy'
import {tryHttpProxyChain, normalizeChain} from './try-http-proxy-chain'

import binary from 'binary'
import domain from 'domain'
import net, {Socket} from 'net'
import {isPrivateAddress} from './utils/is-private-address'

import {ISocksServer, ProxyServers, SetProxyServerOptions, Socks5ProxyOptions} from '../types'
import {HandshakeCompletedHandler} from '../types/handshake'
import {IProxyServer} from '../types/execute-request-through-remote-proxy'
import {
  HttpProxyResolver,
  HttpProxyResolverResult,
} from '../types/execute-request-through-http-proxy'
import {ConnectBufferObject} from '../types/parse-connect-buffer'
import {AddressInfo} from '../types/utils'

const _alwaysTrue = (data: {
  socket: net.Socket
  destinationInfo: AddressInfo
  originInfo: AddressInfo
}) => true

/**
 * The following RFCs may be useful as background:
 *
 * https://www.ietf.org/rfc/rfc1928.txt - NO_AUTH SOCKS5
 * https://www.ietf.org/rfc/rfc1929.txt - USERNAME/PASSWORD SOCKS5
 *
 **/
export class SocksServer implements ISocksServer {
  private _server: net.Server
  private options: Socks5ProxyOptions
  private activeSessions: Socket[] = []
  private destinationSockets: Socket[] = []
  private syncDestinationSocketInterval

  constructor(options: Socks5ProxyOptions) {
    const self = this
    this.options = options || {}

    this.syncDestinationSocketInterval = setInterval(() => {
      this.destinationSockets = this.destinationSockets.filter((dest: Socket) => !dest.destroyed)
    }, 5000)

    this._server = net.createServer((socket) => {
      let clientHandshakeBuffer: Buffer
      socket.on('error', (err) => {
        self.server.emit(EVENTS.ORIGIN_SOCKET_ERROR, {err, socket})
      })
      socket.setKeepAlive(true)

      function _authenticate(buffer: Buffer) {
        authenticate({
          socket,
          buffer,
          options: self.options,
          connect,
          end,
          onAuthenticated: (args) =>
            self.server.emit(EVENTS.AUTHENTICATION, args.username.toString()),
          onFailed: (args, err) =>
            self.server.emit(EVENTS.AUTHENTICATION_ERROR, args.username.toString(), err),
        })
      }
      /**
       * +----+-----+-------+------+----------+----------+
       * |VER | CMD |  RSV  | ATYP | DST.ADDR | DST.PORT |
       * +----+-----+-------+------+----------+----------+
       * | 1  |  1  | X'00' |  1   | Variable |    2     |
       * +----+-----+-------+------+----------+----------+
       *
       * @param {Buffer} buffer - a buffer
       * @returns {undefined}
       **/
      function connect(buffer: Buffer) {
        const onParseCompleted = (args: ConnectBufferObject) => {
          if (args.cmd !== RFC_1928_COMMANDS.CONNECT) {
            // bind and udp associate commands
            return end(RFC_1928_REPLIES.SUCCEEDED, args)
          }

          const connectionFilterDomain = domain.create()
          let connectionFilter = self.options.connectionFilter

          // if no connection filter is provided, stub one
          if (!connectionFilter || typeof connectionFilter !== 'function') {
            connectionFilter = (destination, origin, callback) => setImmediate(callback)
          }

          // capture connection filter errors
          connectionFilterDomain.on('error', (err) => {
            // emit failed destination connection event
            self.server.emit(
              EVENTS.CONNECTION_FILTER,
              // destination
              {
                address: args.dst.addr,
                port: args.dst.port,
              },
              // origin
              {
                address: socket.remoteAddress,
                port: socket.remotePort,
              },
              err,
            )

            // respond with failure
            return end(RFC_1928_REPLIES.CONNECTION_NOT_ALLOWED, args)
            //return end(RFC_1929_REPLIES.CONNECTION_NOT_ALLOWED, args)
          })

          // perform connection
          return connectionFilter(
            // destination
            {
              address: args.dst.addr,
              port: args.dst.port,
            },
            // origin
            {
              address: socket.remoteAddress,
              port: socket.remotePort,
            },
            connectionFilterDomain.intercept(async () => {
              const destinationInfo: AddressInfo = {
                address: args.dst.addr,
                port: args.dst.port,
              }

              const originInfo: AddressInfo = {
                address: socket.remoteAddress,
                port: socket.remotePort,
              }

              // whitelisted addrsses are never routed through a proxy (tunnel) and pass all validations
              // the main use is browser extension communication with internal EC services (e.g. local test runner)
              // const isWhitelisted =
              //   self.options.whitelist?.indexOf(
              //     `${destinationInfo.address}:${destinationInfo.port}`,
              //   ) !== -1
              const isWhitelisted =
                self.options.shouldWhitelist && self.options.shouldWhitelist(destinationInfo)

              const shouldBlockRequest =
                !isWhitelisted &&
                self.options.shouldBlockRequest &&
                self.options.shouldBlockRequest(destinationInfo)

              if (shouldBlockRequest) {
                self.server.emit(EVENTS.ADDRESS_WAS_BLOCKED, {
                  destinationInfo,
                  originInfo,
                })
                return end(RFC_1928_REPLIES.CONNECTION_NOT_ALLOWED, args)
              }

              if (
                !isWhitelisted &&
                shouldBlockRequest !== false &&
                self.options.shouldBlockPrivateAddresses &&
                (await isPrivateAddress(destinationInfo.address!))
              ) {
                self.server.emit(EVENTS.PRIVATE_ADDRESS_WAS_BLOCKED, {
                  destinationInfo,
                  originInfo,
                })
                return end(RFC_1928_REPLIES.CONNECTION_NOT_ALLOWED, args)
              }

              // === CLIENT-SIDE HTTP CONNECT branch ===
              //
              // Consulted BEFORE the SERVER-SIDE SOCKS5-chain branch below.
              // The resolver decides per-request whether this destination is
              // reachable only through a customer corporate HTTP proxy
              // (typical caller: a PAC-file evaluator wrapped in
              // `@applitools/execution-grid-tunnel`).
              //
              // The resolver may return:
              //   - a single `HttpProxyDecision` (single proxy)
              //   - a `ProxyChainEntry[]` (ordered PAC fallback chain;
              //     entries may be `'DIRECT'` to mean PAC's DIRECT directive)
              //   - `null`/`undefined`/`[]` — fall through to the existing
              //     SOCKS5-chain branch, preserving server-side behaviour
              //     (where this resolver is never set).
              //
              // Whitelisted destinations always bypass the resolver — same
              // policy as the existing proxyServers branch.
              if (!isWhitelisted && self.options.httpProxyResolver) {
                let resolverResult: HttpProxyResolverResult
                try {
                  resolverResult = await self.options.httpProxyResolver({
                    socket,
                    destinationInfo,
                    originInfo,
                  })
                } catch (resolverErr) {
                  self.server.emit(EVENTS.HTTP_PROXY_CONNECT_FAILED, {
                    destinationInfo,
                    originInfo,
                    proxyInfo: undefined,
                    error: resolverErr,
                  })
                  return end(RFC_1928_REPLIES.GENERAL_FAILURE, args)
                }
                const chain = normalizeChain(resolverResult)
                if (chain) {
                  const firstEntry = chain[0]
                  const firstProxyInfo =
                    firstEntry === 'DIRECT'
                      ? undefined
                      : {address: firstEntry.address, port: firstEntry.port}
                  self.server.emit(EVENTS.ACCEPT_NEW_REQUEST, {
                    destinationInfo,
                    originInfo,
                    proxyInfo: firstProxyInfo,
                  })

                  // Whole-chain budget timeout (mirrors the existing
                  // per-request timeout). Fires REMOTE_CONNECTION_TIMEOUT_ERROR
                  // observationally — the inner `executeRequest…` modules
                  // have their own tighter timeouts that actually abort.
                  const connectionTimeout = 120000
                  let connectionTimeoutId: NodeJS.Timeout | null = setTimeout(
                    () =>
                      self.server.emit(EVENTS.REMOTE_CONNECTION_TIMEOUT_ERROR, {
                        destinationInfo,
                        originInfo,
                        proxyInfo: firstProxyInfo,
                        timeout: connectionTimeout,
                      }),
                    connectionTimeout,
                  )
                  const clearConnectionTimeout = () => {
                    if (connectionTimeoutId) clearTimeout(connectionTimeoutId)
                    connectionTimeoutId = null
                  }
                  let domainExited = false
                  const exitDomainOnce = () => {
                    if (domainExited) return
                    domainExited = true
                    connectionFilterDomain.exit()
                  }

                  const successfulDestination = await tryHttpProxyChain({
                    originalSocket: socket,
                    chain,
                    addressType: args.atyp,
                    address: args.dst.addr,
                    port: args.dst.port,
                    requestBuffer: args.requestBuffer,
                    destinationInfo,
                    originInfo,
                    end,
                    registerDestination: (d) => self.destinationSockets.push(d),
                    callbacks: {
                      onConnectedToHttpProxy: (entry, dest) => {
                        exitDomainOnce()
                        clearConnectionTimeout()
                        self.server.emit(EVENTS.CONNECTED_TO_HTTP_PROXY, {
                          destinationInfo,
                          originInfo,
                          proxyInfo: {address: entry.address, port: entry.port},
                        })
                      },
                      onConnectedToRemoteAddress: (dest) => {
                        exitDomainOnce()
                        clearConnectionTimeout()
                        self.server.emit(EVENTS.CONNECTED_TO_REMOTE_ADDRESS, {
                          destinationInfo,
                          originInfo,
                          proxyInfo: undefined,
                        })
                      },
                      onProxyResponded: (statusCode, entry) => {
                        self.server.emit(EVENTS.HTTP_PROXY_CONNECT_RESPONDED, {
                          destinationInfo,
                          originInfo,
                          proxyInfo: {address: entry.address, port: entry.port},
                          statusCode,
                        })
                      },
                      onProxyFailed: (rfc1928Reply, entry) => {
                        self.server.emit(EVENTS.HTTP_PROXY_CONNECT_FAILED, {
                          destinationInfo,
                          originInfo,
                          proxyInfo: {address: entry.address, port: entry.port},
                          rfc1928Reply,
                        })
                      },
                      onSocketError: (err, entry, dest) => {
                        exitDomainOnce()
                        const proxyInfo =
                          entry === 'DIRECT'
                            ? undefined
                            : {address: entry.address, port: entry.port}
                        self.server.emit(EVENTS.REMOTE_SOCKET_ERROR, {
                          err,
                          destination: dest,
                          args,
                          originInfo,
                          destinationInfo,
                          proxyInfo,
                        })
                      },
                    },
                  })
                  clearConnectionTimeout()
                  exitDomainOnce()

                  // Register lifecycle cleanup only for the successful
                  // destination (failed attempts' destinations were already
                  // destroyed by the inner functions).
                  if (successfulDestination && !successfulDestination.destroyed) {
                    socket.on('close', () => {
                      successfulDestination.destroy()
                      const idx = self.destinationSockets.indexOf(successfulDestination)
                      if (idx !== -1) self.destinationSockets.splice(idx, 1)
                    })
                  }
                  return
                }
                // chain was null/empty — fall through to existing dispatcher
                // logic (SOCKS5 chain if configured, else DIRECT)
              }
              // === end CLIENT-SIDE HTTP CONNECT branch ===

              let createExecuterAndRegisterToEvents: any = null

              let proxyServer: Partial<IProxyServer> | undefined = undefined

              // whitelisted addresses should never pass through a proxy
              if (!isWhitelisted && self.options.proxyServers?.getAvailableProxy()) {
                proxyServer = {
                  ...self.options.proxyServers.getAvailableProxy(),
                  shouldUseProxy: self.options.proxyServers.shouldUseProxy || _alwaysTrue,
                }
              }

              const shouldUseProxy: boolean =
                proxyServer?.shouldUseProxy!({
                  socket,
                  destinationInfo,
                  originInfo,
                }) || false

              const proxyInfo =
                proxyServer && shouldUseProxy
                  ? {
                      address: proxyServer.address,
                      port: proxyServer.port,
                    }
                  : undefined

              const basicEventData = {
                originInfo,
                destinationInfo,
                proxyInfo,
              }

              self.server.emit(EVENTS.ACCEPT_NEW_REQUEST, basicEventData)

              if (proxyServer && shouldUseProxy) {
                const retries = {handshake: 0, rpcReply: 0}

                proxyServer.onHandshakeCompleted = () => {
                  self.server.emit(EVENTS.REMOTE_PROXY_HANDSHAKE_COMPLETED, basicEventData)
                }

                proxyServer.onHandshakeTimeout = () => {
                  self.server.emit(EVENTS.REMOTE_PROXY_HANDSHAKE_TIMEOUT, {
                    ...basicEventData,
                    timeout: proxyServer!.handshakeTimeout,
                  })

                  retries.handshake++
                  if (retries.handshake <= 3) createExecuterAndRegisterToEvents()
                  else end(RFC_1928_REPLIES.NETWORK_UNREACHABLE, args)
                }

                proxyServer.onRpc1928ReplyTimeout = () => {
                  self.server.emit(EVENTS.REMOTE_PROXY_RFC_1928_REPLY_TIMEOUT, {
                    ...basicEventData,
                    timeout: proxyServer!.rpcReplyTimeout,
                  })

                  retries.rpcReply++
                  if (retries.rpcReply <= 3) createExecuterAndRegisterToEvents()
                  else end(RFC_1928_REPLIES.NETWORK_UNREACHABLE, args)
                }

                proxyServer.onRpc1928Reply = (reply: Buffer) => {
                  self.server.emit(EVENTS.REMOTE_PROXY_RFC_1928_REPLY, {
                    ...basicEventData,
                    reply,
                  })
                }
              }

              createExecuterAndRegisterToEvents = () => {
                const connectionTimeout = 120000
                let connectionTimeoutId: NodeJS.Timeout | null = setTimeout(
                  () =>
                    self.server.emit(EVENTS.REMOTE_CONNECTION_TIMEOUT_ERROR, {
                      ...basicEventData,
                      timeout: connectionTimeout,
                    }),
                  connectionTimeout,
                )

                let destination: net.Socket

                const clearConnectionTimeout = () => {
                  connectionTimeoutId && clearTimeout(connectionTimeoutId)
                  connectionTimeoutId = null
                }

                const onConnect = (destinationSocket: Socket) => {
                  connectionFilterDomain.exit()
                  clearConnectionTimeout()

                  // close destination socket when after original socket was closed
                  socket.on('close', () => {
                    destinationSocket.destroy()
                    self.destinationSockets.splice(
                      self.destinationSockets.indexOf(destinationSocket),
                      1,
                    )
                  })

                  const event = shouldUseProxy
                    ? EVENTS.CONNECTED_TO_REMOTE_PROXY
                    : EVENTS.CONNECTED_TO_REMOTE_ADDRESS
                  self.server.emit(event, {destinationInfo, originInfo, proxyInfo})
                }

                const onError = (err: Error) => {
                  // exit the connection filter domain
                  connectionFilterDomain.exit()

                  self.server.emit(EVENTS.REMOTE_SOCKET_ERROR, {
                    err,
                    destination,
                    args,
                    originInfo,
                    destinationInfo,
                    proxyInfo,
                  })
                }

                if (!shouldUseProxy) {
                  destination = executeRequest({
                    originalSocket: socket,
                    addressType: args.atyp,
                    address: args.dst.addr,
                    port: args.dst.port,
                    requestBuffer: args.requestBuffer,
                    end,
                    onConnect,
                    onError,
                  })
                } else {
                  destination = executeRequestThroughRemoteProxy({
                    originalSocket: socket,
                    clientHandshakeBuffer,
                    proxyServer: proxyServer as IProxyServer,
                    addressType: args.atyp,
                    address: args.dst.addr,
                    port: args.dst.port,
                    requestBuffer: args.requestBuffer,
                    end,
                    onConnect,
                    onError,
                  })
                }

                self.destinationSockets.push(destination)
              }

              createExecuterAndRegisterToEvents()
            }),
          )
        }

        parseConnectBuffer({
          buffer,
          onReqesutBufferParsed: () => self.activeSessions.push(socket),
          onParseCompleted,
          end,
        })
      }

      /**
       * +----+-----+-------+------+----------+----------+
       * |VER | REP |  RSV  | ATYP | BND.ADDR | BND.PORT |
       * +----+-----+-------+------+----------+----------+
       * | 1  |  1  | X'00' |  1   | Variable |    2     |
       * +----+-----+-------+------+----------+----------+
       *
       * @param {Buffer} response - a buffer representing the response
       * @param {object} args - arguments to supply to the proxy end event
       * @returns {undefined}
       **/
      function end(response: number, args: any) {
        endConnection({socket, response, args})
        // indicate end of connection
        self.server.emit(EVENTS.PROXY_END, response, args)
      }

      /**
       * +----+----------+----------+
       * |VER | NMETHODS | METHODS  |
       * +----+----------+----------+
       * | 1  |    1     | 1 to 255 |
       * +----+----------+----------+
       *
       * @param {Buffer} buffer - a buffer
       * @returns {undefined}
       **/

      function _handshake(buffer: Buffer) {
        clientHandshakeBuffer = buffer

        const onHandshakeCompleted: HandshakeCompletedHandler = ({
          clientHandshakeBuffer,
          responseBuffer,
        }) => {
          self.server.emit(EVENTS.HANDSHAKE, {socket, clientHandshakeBuffer, responseBuffer})
        }

        handshake({
          socket,
          buffer,
          options: self.options,
          connect,
          authenticate: _authenticate,
          onHandshakeCompleted,
          end,
        })
      }

      socket.once('data', _handshake)

      // capture socket closure
      socket.once('end', () => {
        // remove the session from currently the active sessions list
        self.activeSessions.splice(self.activeSessions.indexOf(socket), 1)
      })
    })
  }

  public on = (event: string, listener: (...args: any[]) => void) => this.server.on(event, listener)
  public listen = (...args: any[]) => this.server.listen(...args)
  public listenAsync = (port: number, host = '127.0.0.1') => {
    const promise = new Promise<void>((resolve, reject) => {
      try {
        this.server.listen(port, host, resolve)
      } catch (e) {
        reject(e)
      }
    })

    return promise
  }

  public close = (...args: any[]) => {
    this.syncDestinationSocketInterval && clearInterval(this.syncDestinationSocketInterval)
    return this.server.close(...args)
  }
  public getConnections = (cb: (error: Error | null, count: number) => void) =>
    this.server.getConnections(cb)

  // SERVER-SIDE upstream-SOCKS5 chain (used by eg-chromedriver-image for
  // tunnel load-balancing). Not for client-side corporate-proxy routing —
  // see `setHttpProxyResolver` for that.
  public setProxyServer = ({address, port, shouldUseProxy}: SetProxyServerOptions) => {
    const getAvailableProxy = () => ({address, port})
    this.options.proxyServers = {shouldUseProxy, getAvailableProxy}
  }

  public setProxyServers = (proxyServers: ProxyServers) => {
    this.options.proxyServers = proxyServers
  }

  public getProxyServers = () => this.options.proxyServers

  // CLIENT-SIDE per-request HTTP CONNECT proxy resolver (used by
  // execution-grid-tunnel to route AUT traffic through a customer's corporate
  // HTTP proxy, typically driven by a PAC-file evaluator). Returning `null`
  // from the resolver lets the request fall through to the SOCKS5-chain
  // dispatcher branch, preserving compatibility with the server-side use.
  public setHttpProxyResolver = (resolver: HttpProxyResolver) => {
    this.options.httpProxyResolver = resolver
  }

  public set shouldBlockPrivateAddresses(value: boolean) {
    this.options.shouldBlockPrivateAddresses = value
  }

  public get shouldBlockPrivateAddresses() {
    return !!this.options.shouldBlockPrivateAddresses
  }

  public setShouldWhitelist(shouldWhitelist: (data: AddressInfo) => boolean) {
    this.options.shouldWhitelist = shouldWhitelist
  }

  public setShouldBlockRequest(shouldBlockRequest: (data: AddressInfo) => boolean) {
    this.options.shouldBlockRequest = shouldBlockRequest
  }

  public get server() {
    return this._server
  }
}

export function createSocks5ProxyServer(options: Socks5ProxyOptions = {}): SocksServer {
  const socksServer = new SocksServer(options)
  return socksServer
}

export * as EVENTS from './events'

// Public API types for the HTTP CONNECT resolver. Re-exported here so external
// consumers (e.g. `@applitools/execution-grid-tunnel`) can `import {…} from
// '@applitools/eg-socks5-proxy-server'` without reaching into the `types/`
// subpath. The existing SOCKS5-chain types (`ProxyServers`,
// `SetProxyServerOptions`) are not re-exported because they're not part of
// the client-side surface; they're inferable from `SocksServer`'s method
// signatures.
export type {
  HttpProxyDecision,
  HttpProxyResolver,
  HttpProxyResolverContext,
  HttpProxyResolverResult,
  ProxyChainEntry,
} from '../types/execute-request-through-http-proxy'
