// @applitools/eg-socks5-proxy-server/src/try-http-proxy-chain.ts
/**
 * Chain orchestrator for the HTTP CONNECT upstream path.
 *
 * Given an ordered chain of {@link ProxyChainEntry} entries (PAC's
 * `PROXY a:1; PROXY b:2; DIRECT` syntax), tries each in order. On
 * RETRYABLE failure (TCP-level error, proxy 5xx, malformed response) the
 * next entry is attempted; on TERMINAL failure (407 — `CONNECTION_NOT_ALLOWED`)
 * the chain stops and the failure is surfaced to the originating SOCKS5
 * client. On success (any 2xx), the inner `executeRequestThroughHttpProxy`
 * call writes the SOCKS5 SUCCEEDED reply itself and sets up the bidirectional
 * pipe; this function then returns.
 *
 * The `DIRECT` sentinel invokes the same `executeRequest` used by the
 * default dispatcher branch — opening a direct socket to the destination.
 *
 * IMPORTANT — wrapped `end` semantics: the inner execute functions normally
 * call `end(rfcReply, ...)` themselves on failure to write a SOCKS5 FAILURE
 * frame to the originating client. While iterating, we DO NOT want failure
 * frames written until the chain has either succeeded or run out. This
 * orchestrator therefore passes a NO-OP `end` callback to each inner call
 * and calls the real `params.end` once at the right time.
 */
import net, {Socket} from 'net'

import {RFC_1928_REPLIES} from './constants'
import {executeRequest} from './execute-request'
import {executeRequestThroughHttpProxy} from './execute-request-through-http-proxy'

import {
  HttpProxyDecision,
  ProxyChainEntry,
} from '../types/execute-request-through-http-proxy'
import {AddressInfo} from '../types/utils'

// RFC 1928 reply codes we consider RETRYABLE — i.e. the chain should move to
// the next entry. These correspond to transport/proxy availability problems.
// Anything else (notably 407 → CONNECTION_NOT_ALLOWED) is TERMINAL: it tells
// us "this specific proxy rejected the request for a reason that won't be
// fixed by trying a different proxy."
const RETRYABLE_RFC_CODES = new Set<number>([
  RFC_1928_REPLIES.NETWORK_UNREACHABLE, // 502/503 or convertSocketErrorToRpcReply default
  RFC_1928_REPLIES.HOST_UNREACHABLE, // EADDRNOTAVAIL
  RFC_1928_REPLIES.CONNECTION_REFUSED, // ECONNREFUSED or proxy hang-up
  RFC_1928_REPLIES.TTL_EXPIRED, // 504, header timeout
  RFC_1928_REPLIES.GENERAL_FAILURE, // malformed status line, header overflow
])

export interface ProxyChainCallbacks {
  /** Fires for each non-DIRECT attempt once TCP/TLS to the proxy is up. */
  onConnectedToHttpProxy?: (entry: HttpProxyDecision, destination: Socket) => void
  /** Fires for the DIRECT attempt (if present) once the socket to the destination is up. */
  onConnectedToRemoteAddress?: (destination: Socket) => void
  /** Fires per attempt when the proxy returns an HTTP status line. */
  onProxyResponded?: (statusCode: number, entry: HttpProxyDecision) => void
  /** Fires per attempt on failure (transport or non-2xx response). */
  onProxyFailed?: (rfc1928Reply: number, entry: HttpProxyDecision) => void
  /** Fires on any socket error during an attempt (post-tunnel pipe errors included). */
  onSocketError?: (
    err: Error,
    entry: ProxyChainEntry,
    destination?: Socket,
  ) => void
}

export interface TryHttpProxyChainParameters {
  originalSocket: Socket
  chain: ProxyChainEntry[]
  addressType: number
  address: string
  port: number
  requestBuffer: Buffer
  /**
   * Writes the final SOCKS5 reply to `originalSocket`. Called by the
   * orchestrator at most once: either implicitly by the inner functions on
   * success (they write a SOCKS5 SUCCEEDED themselves and pipe), or
   * explicitly here on terminal/exhausted failure with the last RFC 1928
   * reply code.
   */
  end: (rfcReply: number, args: {requestBuffer: Buffer}) => void
  callbacks: ProxyChainCallbacks
  /**
   * Per-attempt destination registration. Used by the dispatcher to track
   * destination sockets for leak detection. Each attempt registers its own
   * destination; failed-attempt destinations are destroyed by the inner
   * functions and will be filtered out by the periodic sweep.
   */
  registerDestination?: (destination: Socket) => void
  destinationInfo: AddressInfo
  originInfo: AddressInfo
}

interface AttemptOutcome {
  kind: 'success' | 'retryable' | 'terminal'
  destination: Socket
  rfc: number
}

/**
 * Iterates the chain. Returns the destination socket of the successful
 * attempt (for the dispatcher to attach lifecycle handlers), or
 * `undefined` if the chain exhausted without success.
 *
 * Side effects (which the dispatcher relies on):
 *   - On success: `executeRequestThroughHttpProxy` (or `executeRequest`)
 *     has already written the SOCKS5 SUCCEEDED reply to `originalSocket`
 *     and wired the bidirectional pipe.
 *   - On terminal/exhausted failure: `params.end(rfc, ...)` has been
 *     called with the last RFC 1928 reply code.
 *   - For every attempt: callbacks fire as documented.
 */
export async function tryHttpProxyChain(
  params: TryHttpProxyChainParameters,
): Promise<Socket | undefined> {
  let lastRfcReply = RFC_1928_REPLIES.GENERAL_FAILURE
  let lastDestination: Socket | undefined

  for (let i = 0; i < params.chain.length; i++) {
    const entry = params.chain[i]

    if (entry === 'DIRECT') {
      const outcome = await runDirectEntry(params)
      lastRfcReply = outcome.rfc
      lastDestination = outcome.destination
      if (outcome.kind === 'success') return outcome.destination
      // DIRECT failure is terminal — there's nothing to fall back to.
      params.end(outcome.rfc, {requestBuffer: params.requestBuffer})
      return outcome.destination
    }

    const outcome = await runProxyEntry(entry, params)
    lastRfcReply = outcome.rfc
    lastDestination = outcome.destination
    if (outcome.kind === 'success') return outcome.destination
    if (outcome.kind === 'terminal') {
      params.end(outcome.rfc, {requestBuffer: params.requestBuffer})
      return outcome.destination
    }
    // retryable — try next entry
  }

  // Chain exhausted with no success — surface the last failure.
  params.end(lastRfcReply, {requestBuffer: params.requestBuffer})
  return lastDestination
}

function runProxyEntry(
  entry: HttpProxyDecision,
  params: TryHttpProxyChainParameters,
): Promise<AttemptOutcome> {
  return new Promise((resolve) => {
    let settled = false
    // We capture the destination synchronously below — but the closure here
    // needs forward-declared visibility so the callbacks can reference it.
    let destinationRef: Socket | undefined

    const settle = (outcome: AttemptOutcome) => {
      if (settled) return
      settled = true
      resolve(outcome)
    }

    const destination = executeRequestThroughHttpProxy({
      originalSocket: params.originalSocket,
      proxy: entry,
      addressType: params.addressType,
      address: params.address,
      port: params.port,
      requestBuffer: params.requestBuffer,
      onConnect: (sock) => {
        // TCP/TLS to the proxy is up; surface the lifecycle event so the
        // dispatcher can clear timeouts and register cleanup.
        params.callbacks.onConnectedToHttpProxy?.(entry, sock)
      },
      onProxyResponded: (statusCode) => {
        params.callbacks.onProxyResponded?.(statusCode, entry)
        if (statusCode >= 200 && statusCode < 300) {
          // The inner function will write the SOCKS5 SUCCEEDED frame and
          // wire the pipe in the same synchronous block. Resolve as
          // success — the pipe will be live by the next tick.
          settle({kind: 'success', destination: destinationRef!, rfc: statusCode})
        }
      },
      onProxyFailed: (rfc) => {
        params.callbacks.onProxyFailed?.(rfc, entry)
        const isRetryable = RETRYABLE_RFC_CODES.has(rfc)
        settle({
          kind: isRetryable ? 'retryable' : 'terminal',
          destination: destinationRef!,
          rfc,
        })
      },
      // NO-OP `end` while iterating. The orchestrator will call the real
      // `params.end` once chain-level outcome is decided.
      end: () => {
        /* deliberately empty — see file-level comment */
      },
      onError: (err) => {
        params.callbacks.onSocketError?.(err, entry, destinationRef)
      },
    })
    destinationRef = destination
    params.registerDestination?.(destination)
  })
}

function runDirectEntry(
  params: TryHttpProxyChainParameters,
): Promise<AttemptOutcome> {
  return new Promise((resolve) => {
    let settled = false
    let destinationRef: Socket | undefined

    const settle = (outcome: AttemptOutcome) => {
      if (settled) return
      settled = true
      resolve(outcome)
    }

    const destination = executeRequest({
      originalSocket: params.originalSocket,
      addressType: params.addressType,
      address: params.address,
      port: params.port,
      requestBuffer: params.requestBuffer,
      end: (rfc) => {
        // DIRECT path failed before establishing — surface as terminal.
        settle({kind: 'terminal', destination: destinationRef!, rfc})
      },
      onConnect: (sock) => {
        // `executeRequest` writes the SOCKS5 SUCCEEDED reply and sets up
        // the pipe BEFORE invoking onConnect. So we can settle as success
        // right here.
        params.callbacks.onConnectedToRemoteAddress?.(sock)
        settle({
          kind: 'success',
          destination: destinationRef!,
          rfc: RFC_1928_REPLIES.SUCCEEDED,
        })
      },
      onError: (err) => {
        params.callbacks.onSocketError?.(err, 'DIRECT', destinationRef)
      },
    })
    destinationRef = destination
    params.registerDestination?.(destination)
  })
}

/**
 * Normalize a resolver result into a chain.
 *   - `null` / `undefined` / `[]` → `null` (caller falls through)
 *   - single decision → `[decision]`
 *   - array → used as-is
 */
export function normalizeChain(
  result:
    | HttpProxyDecision
    | ProxyChainEntry[]
    | null
    | undefined,
): ProxyChainEntry[] | null {
  if (result == null) return null
  if (Array.isArray(result)) return result.length === 0 ? null : result
  return [result]
}
