// @applitools/dom-shared/plugins/sandbox.js
const path = require('path')

function sandbox({plugins = []} = {}) {
  const originalOutputs = {}
  return {
    name: 'sandbox',
    outputOptions(output) {
      originalOutputs[output.file || output.name] = output
      return {...output, format: 'iife'}
    },
    async generateBundle(output, bundle) {
      const rollup = require('rollup')
      const resolve = require('@rollup/plugin-node-resolve')
      const commonjs = require('@rollup/plugin-commonjs')
      const replace = require('@rollup/plugin-replace')
      const babel = require('@rollup/plugin-babel')

      const originalOutput = originalOutputs[output.file || output.name]
      Object.assign(output, originalOutput)

      const [file] = Object.values(bundle)
      const build = await rollup.rollup({
        input: path.resolve(__dirname, './sandbox-wrapper.ts'),
        plugins: [
          resolve({browser: true}),
          commonjs({ignoreGlobal: true}),
          babel.default({
            babelHelpers: 'inline',
            extensions: ['.ts'],
            presets: [['@babel/preset-typescript']],
          }),
          replace({
            delimiters: ['__', '__'],
            values: {
              IIFE: `function(document, window, cleanWindow) {return ${file.code.replace(/^var \w+\s*=\s*/, '')}}`,
            },
            preventAssignment: true,
          }),
          ...plugins,
        ],
      })
      const wrapped = await build.generate(originalOutput)
      file.code = wrapped.output[0].code
    },
  }
}

module.exports = sandbox
