// @applitools/logger/dist/spy-logger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeSpyLogger = void 0;
/* eslint no-console: off */
const logger_1 = require("./logger");
const utils = __importStar(require("@applitools/utils"));
// @ts-ignore
const promises_1 = require("node:timers/promises");
function makeSpyLogger({ level, label } = {}) {
    const calls = [];
    const waiter = makeWaiter();
    let loggerExtension = undefined;
    // When SHOW_SPY_LOGS (APPLITOOLS_SHOW_SPY_LOGS) is enabled, mirror captured calls to the
    // console so developers can see what the SDK is doing while spy assertions still work.
    // Pure spies otherwise produce zero visible output, which makes debugging hangs and
    // failures slow. A dedicated env var (rather than reusing APPLITOOLS_SHOW_LOGS) keeps
    // this opt-in for spec-debugging without also flipping every real logger in the process
    // to verbose mode. Suppressed when CI is set: the visibility is a local-dev aid, and
    // mirroring in CI just pollutes build logs without helping anyone.
    const mirrorToConsole = utils.general.getEnvValue('SHOW_SPY_LOGS', 'boolean') && !process.env.CI;
    const prefix = label ? `[spy:${label}]` : '[spy]';
    const logger = (0, logger_1.makeLogger)({
        handler: {
            log(message) {
                calls.push({ level: 'info', message });
                loggerExtension === null || loggerExtension === void 0 ? void 0 : loggerExtension.log(message);
                if (mirrorToConsole)
                    console.log(`${prefix} [INFO ] ${message}`);
                waiter.notifyAll();
            },
            error(message) {
                calls.push({ level: 'error', message });
                loggerExtension === null || loggerExtension === void 0 ? void 0 : loggerExtension.error(message);
                if (mirrorToConsole)
                    console.error(`${prefix} [ERROR] ${message}`);
                waiter.notifyAll();
            },
            warn(message) {
                calls.push({ level: 'warn', message });
                loggerExtension === null || loggerExtension === void 0 ? void 0 : loggerExtension.warn(message);
                if (mirrorToConsole)
                    console.warn(`${prefix} [WARN ] ${message}`);
                waiter.notifyAll();
            },
            fatal(message) {
                calls.push({ level: 'fatal', message });
                loggerExtension === null || loggerExtension === void 0 ? void 0 : loggerExtension.fatal(message);
                if (mirrorToConsole)
                    console.error(`${prefix} [FATAL] ${message}`);
                waiter.notifyAll();
            },
        },
        level,
    });
    const spyLogger = {
        ...logger,
        calls,
        waitForNextCall(timeout = 5000) {
            const start = Date.now();
            return waiter.wait(start + timeout, `Timeout waiting for the next logger call. Current calls: ${calls.length}. Timeout: ${timeout}ms`);
        },
        async waitForLength(length, timeout = 5000) {
            const start = Date.now();
            while (calls.length < length) {
                await waiter.wait(start + timeout, `Timeout waiting for logger to have ${length} calls reached. Current calls: ${calls.length}. Timeout: ${timeout}ms`);
            }
        },
        async waitForLog(message, { level, timeout = 5000 } = {}) {
            const start = Date.now();
            while (!calls.some(call => call.message.includes(message) && (!level || call.level === level))) {
                await waiter.wait(start + timeout, `Timeout waiting for logger to have message "${message}". Timeout: ${timeout}ms`);
            }
        },
        extend: (_optionsOrLogger, _extraOptions) => {
            return spyLogger;
        },
        wrap: (anotherLogger) => {
            loggerExtension = anotherLogger;
            return logger;
        },
        get isSpyLogger() {
            return true;
        },
    };
    return spyLogger;
}
exports.makeSpyLogger = makeSpyLogger;
function makeWaiter() {
    const waiters = [];
    return {
        wait(deadline, errorMessage) {
            const timeLeft = deadline - Date.now();
            const abortController = new AbortController();
            if (timeLeft > 0) {
                return Promise.race([
                    new Promise(res => waiters.push(res)),
                    (0, promises_1.setTimeout)(timeLeft, new Error(errorMessage !== null && errorMessage !== void 0 ? errorMessage : `Timeout waiting for logger`), {
                        signal: abortController.signal,
                    }),
                ]).then(error => {
                    abortController.abort();
                    if (error instanceof Error) {
                        throw error;
                    }
                });
            }
            throw new Error(errorMessage !== null && errorMessage !== void 0 ? errorMessage : `Timeout waiting for logger`);
        },
        notifyAll() {
            waiters.splice(0).forEach(res => res());
        },
    };
}
