// @applitools/core/dist/ufg/create-render-target-from-snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRenderTargetFromSnapshot = void 0;
async function createRenderTargetFromSnapshot({ ufgClient, snapshot, logger, urlWithCredentials, }) {
    return await ufgClient.createRenderTarget({
        snapshot: snapshot.target,
        settings: {
            url: urlWithCredentials !== null && urlWithCredentials !== void 0 ? urlWithCredentials : snapshot.url,
            environment: snapshot.environment,
            cookies: snapshot.target.cookies,
            headers: {
                Referer: snapshot.url,
                'User-Agent': snapshot.userAgent,
                ...snapshot.settings.headers,
            },
            proxy: snapshot.account.eyesServer.proxy,
            autProxy: snapshot.settings.autProxy,
            skipRootHtmlResource: snapshot.settings.skipRootHtmlResource,
            httpVersion: snapshot.settings.httpVersion,
        },
        logger,
    });
}
exports.createRenderTargetFromSnapshot = createRenderTargetFromSnapshot;
