// @applitools/core/dist/ufg/to-old-infra-environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toOldInfraEnvironment = void 0;
const latestToOneBack = {
    chrome: 'chrome-one-version-back',
    firefox: 'firefox-one-version-back',
    safari: 'safari-one-version-back',
    edgechromium: 'edgechromium-one-version-back',
};
/**
 * Returns the same UFG environment retargeted at the previous browser/OS version,
 * or `null` if the input is not at a latest-alias environment (already one-version-back,
 * pinned to a specific version, or an unsupported browser).
 */
function toOldInfraEnvironment(env) {
    if (!isLatest(env))
        return null;
    if ('iosDeviceInfo' in env)
        return { ...env, iosDeviceInfo: { ...env.iosDeviceInfo, version: 'latest-1' } };
    if ('androidDeviceInfo' in env)
        return { ...env, androidDeviceInfo: { ...env.androidDeviceInfo, version: 'latest-1' } };
    if ('chromeEmulationInfo' in env)
        return { ...env, name: 'chrome-one-version-back' };
    return { ...env, name: latestToOneBack[env.name] };
}
exports.toOldInfraEnvironment = toOldInfraEnvironment;
function isLatest(env) {
    if ('iosDeviceInfo' in env)
        return env.iosDeviceInfo.version !== 'latest-1';
    if ('androidDeviceInfo' in env)
        return env.androidDeviceInfo.version !== 'latest-1';
    if ('chromeEmulationInfo' in env)
        return !env.name || env.name === 'chrome';
    return !!env.name && env.name in latestToOneBack;
}
